File: lib/AbstractLayer/AbstractLayer.cpp
//
//  AbstractLayer.cpp
//  lib
//
//  Created by mengxk on 2018/12/28.
//  Copyright © 2018 Elastos. All rights reserved.
//

#include "AbstractLayer.hpp"

#include <Elastos.SDK.Wallet.C/IdentityManager.h>

char* copyStr2Buf(std::string orgStr){
    
    char* ret = nullptr;
    int nLen = (int)orgStr.length()+1;
    ret = (char*)malloc(nLen);
    memset(ret, 0, nLen);
    orgStr.copy(ret,nLen,0);
    
    return ret;
}

void AbstractLayer_IdentityManager_FreeBuf(void* buf)
{
    free(buf);
}

char* AbstractLayer_IdentityManager_GetMnemonic(const char* language, const char* words)
{
    char* ret = nullptr;
    //printf("start AbstractLayer.cpp\n");
    std::string mnemonic = elastos::IdentityManager::GetMnemonic(language, words);
    ret = copyStr2Buf(mnemonic);
    
    return ret;
}

char* AbstractLayer_IdentityManager_GetSeed(const char*mnemonic, const char* language, const char* words,const char* mnemonicPassword)
{
    
    char* ret = nullptr;
    //printf("start AbstractLayer.cpp\n");
    std::string seed = elastos::IdentityManager::GetSeed(mnemonic, language, words,mnemonicPassword);
    ret = copyStr2Buf(seed);
    
    return ret;
}

char* AbstractLayer_IdentityManager_GetSignInfo(const char* path,const char* url,const char* seed,const char* jsonData){
    
    char* ret = nullptr;
    
    //create identity object
    std::shared_ptr<elastos::Identity> Identity;
    int nRet = elastos::IdentityManager::CreateIdentity(path, &Identity);
    if(nRet != 0){
        return nullptr;
    }
    
    //create Single AddressWallet
    //const std::shared_ptr<elastos::BlockChainNode> Node = (const std::shared_ptr<elastos::BlockChainNode>)new elastos::BlockChainNode(url);
    const std::shared_ptr<elastos::BlockChainNode> Node(new elastos::BlockChainNode(url));
    std::shared_ptr<elastos::HDWallet> HDWallet;
    Identity->CreateSingleAddressWallet(seed, Node, &HDWallet);
    if(nRet != 0){
        return nullptr;
    }
    
    //create DID manager
    std::shared_ptr<elastos::DidManager> DidManager;
    nRet = Identity->CreateDidManager(seed, &DidManager);
    if(nRet != 0){
        return nullptr;
    }
    
    //create DID
    std::shared_ptr<elastos::Did> Did;
    nRet = DidManager->CreateDid(0, &Did);
    if(nRet != 0){
        return nullptr;
    }
    
    std::string info = Did->SignInfo(seed, jsonData);
    ret = copyStr2Buf(info);
    
    return ret;
}


char* AbstractLayer_IdentityManager_GetInfo(const char* path,const char* seed,const char* key){
    
    char* ret = nullptr;
    
    //create identity object
    std::shared_ptr<elastos::Identity> Identity;
    int nRet = elastos::IdentityManager::CreateIdentity(path, &Identity);
    if(nRet != 0){
        return nullptr;
    }
    
    //create DID manager
    std::shared_ptr<elastos::DidManager> DidManager;
    nRet = Identity->CreateDidManager(seed, &DidManager);
    if(nRet != 0){
        return nullptr;
    }
    
    //create DID
    std::shared_ptr<elastos::Did> Did;
    nRet = DidManager->CreateDid(0, &Did);
    if(nRet != 0){
        return nullptr;
    }
    
    nRet = Did->SyncInfo();
    if(nRet < 0){
        return nullptr;
    }
    
    
    std::string info = Did->GetInfo(key);
    ret = copyStr2Buf(info);
    
    return ret;
}

File: lib/AbstractLayer/AbstractLayer.swift
//
//  AbstractLayer.swift
//  lib
//
//  Created by mengxk on 2018/12/28.
//  Copyright © 2018 Elastos. All rights reserved.
//

import Foundation

final class AbstractLayer {
    private init() {}
  
//    open class IdentityManager {
//        
//        public static func GetMnemonic(language: String, words: String?) -> String? {
//            
//            //print("start Abstract.swift GetMnemonic")
//            
//            let languagePtr = String.ToUnsafeMutablePointer(data: language)
//            let wordsPtr = String.ToUnsafeMutablePointer(data: words)
//            
//            let mnemonicPtr = AbstractLayer_IdentityManager_GetMnemonic(languagePtr, wordsPtr)
//            //print(mnemonicPtr)
//            let mnemonic = String.FromUnsafeMutablePointer(data: mnemonicPtr)
//            //print("end Abstract.swift GetMnemonic")
//            
//            AbstractLayer_IdentityManager_FreeBuf(mnemonicPtr)
//            
//            return mnemonic
//        }
//        
//        public static func GetSeed(mnemonic: String,language: String, words: String?,mnemonicPassword: String?) -> String? {
//            
//            //print("start Abstract.swift GetMnemonic")
//            let mnemonicPtr = String.ToUnsafeMutablePointer(data: mnemonic)
//            let languagePtr = String.ToUnsafeMutablePointer(data: language)
//            let wordsPtr = String.ToUnsafeMutablePointer(data: words)
//            let mnemonicPasswordPtr = String.ToUnsafeMutablePointer(data: mnemonicPassword)
//            
//            let seedPtr = AbstractLayer_IdentityManager_GetSeed(mnemonicPtr, languagePtr, wordsPtr, mnemonicPasswordPtr)
//            
//            //print(mnemonicPtr)
//            let seed = String.FromUnsafeMutablePointer(data: seedPtr)
//            //print("end Abstract.swift GetMnemonic")
//            AbstractLayer_IdentityManager_FreeBuf(seedPtr)
//            
//            return seed
//        }
//        
//        
//        public static func GetSignInfo(path:String,url:String,seed:String,jsonData:String) -> String?{
//            
//            //
//            let pathPtr = String.ToUnsafeMutablePointer(data: path)
//            let urlPtr = String.ToUnsafeMutablePointer(data: url)
//            let seedPtr = String.ToUnsafeMutablePointer(data: seed)
//            let jsonDataPtr = String.ToUnsafeMutablePointer(data: jsonData)
//            
//            let signInfoPtr = AbstractLayer_IdentityManager_GetSignInfo(pathPtr,urlPtr,seedPtr,jsonDataPtr)
//            let signInfo = String.FromUnsafeMutablePointer(data: signInfoPtr)
//            
//            AbstractLayer_IdentityManager_FreeBuf(signInfoPtr)
//            
//            return signInfo;
//            
//        }
//        
//        public static func GetInfo(path:String,seed:String,key:String) -> String?{
//            
//            //
//            let pathPtr = String.ToUnsafeMutablePointer(data: path)
//            let seedPtr = String.ToUnsafeMutablePointer(data: seed)
//            let keyPtr = String.ToUnsafeMutablePointer(data: key)
//            
//            let infoPtr = AbstractLayer_IdentityManager_GetInfo(pathPtr,seedPtr,keyPtr)
//            let info = String.FromUnsafeMutablePointer(data: infoPtr)
//            
//            AbstractLayer_IdentityManager_FreeBuf(infoPtr)
//            
//            return info;
//            
//        }
//        
//        
//    }
//    
//    
}

